### PR TITLE
Merge v7 Spectacle into main

### PR DESCRIPTION
**Pull request by carlos-kelly**
Every commit inside v7 has been PR'd individually into the `main-7` branch but this will set the `main` branch to the v7 codebase.

### CODE REVIEW

**Review by ryan-roemer — CHANGES_REQUESTED** [review]
Some mostly minor things:

- [x] `docs` aren't passing CI. `cd docs && yarn && yarn check-ci` fails presumably because of upgraded eslint config in root that we use in `docs/`
- [ ] `yarn && yarn build-one-page` produce git changes in root.
- [ ] Animated bullets aren't working in spectacle/examples/one-page.html?immediate=true&slide=3&slideElement=1

File: src/utils/remark-rehype-presenter-notes.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+export default function remarkRehypePresenterNotes(noteCallback) {
+  function transformZoneNote(start, nodes) {
+    noteCallback(...nodes);
+    return [];
+  }
+
+  function transformLineNote(node, index, parent) {
+    if (node.children.length === 0) return;
+    if (node.children[0].type !== 'text') return;
```
Comment (upatel32) [inline src/utils/remark-rehype-presenter-notes.js:13]: nit: not important, worth combining these two into one statement?

File: src/hooks/use-slides.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    if (!slideContainer) return;
+    const slides = slideContainer.getElementsByClassName(
+      PLACEHOLDER_CLASS_NAME
+    );
+
+    const nextSlideIds = [];
+    for (const placeholderNode of slides) {
+      const { slideId } = placeholderNode.dataset;
+      nextSlideIds.push(slideId);
+    }
```
Comment (upatel32) [inline src/hooks/use-slides.js:25]: nit: could do

```
const nextSlideIds = slides.map(slide => slide.dataset.slideId);
```

File: src/hooks/use-aspect-ratio-fitting.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    },
+    [targetWidth, targetHeight]
+  );
+
+  // recalculate sizes on the initial pass, and each time the target size
+  // changes. (our measurements aren't as accurate as `useResizeObserver`, but
+  // we only need to get them close because it'll do them again anyways.)
+  React.useEffect(() => {
+    if (!containerRef) return;
+    if (!containerRef.current) return;
```
Comment (upatel32) [inline src/hooks/use-aspect-ratio-fitting.js:50]: same as before, could combine these

File: src/hooks/use-aspect-ratio-fitting.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        : containerWidth / targetWidth;
+
+      const scaledWidth = targetWidth * scaleFactor;
+      const scaledHeight = targetHeight * scaleFactor;
+
+      let x0 = 0;
+      if (useVertical) {
+        x0 = 0.5 * (containerWidth - scaledWidth);
+        x0 /= 1 - scaleFactor;
+      }
```
Comment (upatel32) [inline src/hooks/use-aspect-ratio-fitting.js:31]: I was thinking if we wanted to use const we could use a ternary operator here, but I think the way it currently is helps readability

File: src/components/code-pane.js [hunk 5]
```diff
@@ -1,172 +1,224 @@
...
+      return { opacity: 0.5 };
     }
+  }
+
+  const [from, to] = activeRange;
+  if (from <= lineNumber && lineNumber <= to) {
+    return { opacity: 1 };
+  } else {
+    return { opacity: 0.5 };
+  }
```
Comment (upatel32) [inline src/components/code-pane.js:79]: this could be ternary

**Review by upatel32 — APPROVED** [review]
Everything looks good! I left a few comments that were all mainly about syntactic sugar. Feel free to ignore them!

**Review by ryan-roemer — APPROVED** [review]
Works great! 🎉

**Pull request merged**